cowdao_cowpy/post_process.py

```diff
@@ -0,0 +1,120 @@
+"""
+Post-process OpenAPI spec to update model files with nullable fields.
+"""
+
+import sys
+import ast
+import yaml
+from pathlib import Path
+from typing import Dict, Set
+
+
+def extract_nullable_fields(openapi_spec_path: str) -> Dict[str, Set[str]]:
+    """Parse OpenAPI spec and extract nullable fields for each schema."""
+    with open(openapi_spec_path, "r") as f:
+        spec = yaml.safe_load(f)
+
+    nullable_fields_by_class = {}
+
+    components = spec.get("components", {}).get("schemas", {})
+    for class_name, schema in components.items():
+        if schema.get("type") != "object":
+            continue
+        properties = schema.get("properties", {})
+        for prop_name, prop_attrs in properties.items():
+            if prop_attrs.get("nullable") is True:
+                if "allOf" in prop_attrs:
+                    nullable_fields_by_class.setdefault(class_name, set()).add(
+                        prop_name
+                    )
+
+    return nullable_fields_by_class
+
+
+def update_ast_model_file(
+    model_file_path: str, nullable_fields_by_class: Dict[str, Set[str]]
+) -> None:
+    """Update AST of the model file to make fields Optional if marked nullable in OpenAPI."""
+    code = Path(model_file_path).read_text()
+    tree = ast.parse(code)
+
+    class NullableFieldRewriter(ast.NodeTransformer):
+        def visit_ClassDef(self, node: ast.ClassDef) -> ast.ClassDef:
+            nullable_fields = nullable_fields_by_class.get(node.name, set())
+
+            for stmt in node.body:
+                if not isinstance(stmt, ast.AnnAssign):
+                    continue
+                if not isinstance(stmt.target, ast.Name):
+                    continue
+
+                field_name = stmt.target.id
+                if field_name not in nullable_fields:
+                    continue
+
+                # Only proceed if we're going to modify both type and value
+                if not isinstance(stmt.annotation, ast.Name):
+                    continue
+                if not (
+                    isinstance(stmt.value, ast.Call)
+                    and isinstance(stmt.value.func, ast.Name)
+                    and stmt.value.func.id == "Field"
+                ):
+                    continue
+                if not stmt.value.args:
+                    continue
+                if not (
+                    isinstance(stmt.value.args[0], ast.Constant)
+                    and stmt.value.args[0].value is Ellipsis
+                ):
+                    continue
+
+                print(
```

Comment:
minor, but `#would` use a logger instead of prints

Reply:
hahahaha.
Fine.

Reply:
https://github.com/cowdao-grants/cow-py/pull/59/commits/47e7660614e286c2bb768d032de0f4b551b9bbcc
Actioned here.